### DISCUSSION TITLE
Error raised when an app is terminated with an empty ``SelectionList``.

### DISCUSSION THREAD

**Discussion by YooSunYoung** [Textualize/textual#3559] [Q&A] [answered]
Hi!

I was writing a simple tool to add/delete kafka topics with textual.
Whenever I exit the app after deleting all ``Selection`` options in the ``SelectionList``, it raised ``OptionDoesNotExist`` error that I don't understand...

Can you please help me...?

## Error Message
<details>
<summary>Full Error Message</summary>
    Traceback (most recent call last):
      File ".tox/dev/lib/python3.10/site-packages/textual/message_pump.py", line 536, in _process_messages_loop
        await self._dispatch_message(message)
      File ".tox/dev/lib/python3.10/site-packages/textual/message_pump.py", line 599, in _dispatch_message
        await self._on_message(message)
      File ".tox/dev/lib/python3.10/site-packages/textual/message_pump.py", line 687, in _on_message
        await invoke(method, message)
      File ".tox/dev/lib/python3.10/site-packages/textual/_callback.py", line 74, in invoke
        return await _invoke(callback, *params)
      File ".tox/dev/lib/python3.10/site-packages/textual/_callback.py", line 34, in _invoke
        result = callback(*params[:parameter_count])
      File ".tox/dev/lib/python3.10/site-packages/textual/widgets/_selection_list.py", line 579, in _on_option_list_option_highlighted
        self.post_message(self.SelectionHighlighted(self, event.option_index))
      File ".tox/dev/lib/python3.10/site-packages/textual/widgets/_selection_list.py", line 172, in __init__
        ] = selection_list.get_option_at_index(index)
      File ".tox/dev/lib/python3.10/site-packages/textual/widgets/_selection_list.py", line 603, in get_option_at_index
        return cast("Selection[SelectionType]", super().get_option_at_index(index))
      File "/.tox/dev/lib/python3.10/site-packages/textual/widgets/_option_list.py", line 836, in get_option_at_index
        raise OptionDoesNotExist(
    textual.widgets._option_list.OptionDoesNotExist: There is no option with an index of 0
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "~/anaconda3/envs/benchmark/lib/python3.10/runpy.py", line 196, in _run_module_as_main
        return _run_code(code, main_globals, None,
      File "~/anaconda3/envs/benchmark/lib/python3.10/runpy.py", line 86, in _run_code
        exec(code, run_globals)
      File "./tests/executables/empty_selection_list_error.py", line 162, in <module>
        app.run()
      File ".tox/dev/lib/python3.10/site-packages/textual/app.py", line 1400, in run
        asyncio.run(run_app())
      File "~/anaconda3/envs/benchmark/lib/python3.10/asyncio/runners.py", line 44, in run
        return loop.run_until_complete(main)
      File "~/anaconda3/envs/benchmark/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
        return future.result()
      File ".tox/dev/lib/python3.10/site-packages/textual/app.py", line 1389, in run_app
        await self.run_async(
      File ".tox/dev/lib/python3.10/site-packages/textual/app.py", line 1361, in run_async
        await app._shutdown()
      File ".tox/dev/lib/python3.10/site-packages/textual/app.py", line 2424, in _shutdown
        await self._close_all()
      File ".tox/dev/lib/python3.10/site-packages/textual/app.py", line 2416, in _close_all
        await child._close_messages()
      File ".tox/dev/lib/python3.10/site-packages/textual/message_pump.py", line 462, in _close_messages
        await self._task
      File ".tox/dev/lib/python3.10/site-packages/textual/message_pump.py", line 481, in _process_messages
        await self._process_messages_loop()
      File ".tox/dev/lib/python3.10/site-packages/textual/message_pump.py", line 541, in _process_messages_loop
        self.app._handle_exception(error)
      File ".tox/dev/lib/python3.10/site-packages/textual/app.py", line 2064, in _handle_exception
        self._fatal_error()
      File ".tox/dev/lib/python3.10/site-packages/textual/app.py", line 2069, in _fatal_error
        traceback = Traceback(
      File ".tox/dev/lib/python3.10/site-packages/rich/traceback.py", line 264, in __init__
        trace = self.extract(
      File ".tox/dev/lib/python3.10/site-packages/rich/traceback.py", line 449, in extract
        locals={
      File ".tox/dev/lib/python3.10/site-packages/rich/traceback.py", line 450, in <dictcomp>
        key: pretty.traverse(
      File ".tox/dev/lib/python3.10/site-packages/rich/pretty.py", line 853, in traverse
        node = _traverse(_object, root=True)
      File ".tox/dev/lib/python3.10/site-packages/rich/pretty.py", line 648, in _traverse
        args = list(iter_rich_args(rich_repr_result))
      File ".tox/dev/lib/python3.10/site-packages/rich/pretty.py", line 615, in iter_rich_args
        for arg in rich_args:
      File ".tox/dev/lib/python3.10/site-packages/textual/widgets/_selection_list.py", line 189, in __rich_repr__
        yield "selection", self.selection
    AttributeError: 'SelectionHighlighted' object has no attribute 'selection'

</details>

## Reproduce Error
- python version: 3.10
- textual version: 0.40.0

Here is the code snippet of very simplified app that can reproduce the error:
<details>
<summary> empty_selection_list_error.py </summary>

    from textual.app import App, ComposeResult, Screen
    from textual.containers import Container, Horizontal, VerticalScroll
    from textual.widgets import Button, Header, Pretty, SelectionList, Static
    from textual.widgets.selection_list import Selection
    
    class FakeServer:
        def __init__(self) -> None:
            self._options = dict()
        
        @property
        def options(self) -> dict:
            return self._options
    
        def populate_option(self):
            import uuid
            new_opt = uuid.uuid4().hex
            self._options.setdefault(new_opt, new_opt)
        
        def remove_option(self, opt_name) -> None:
            self._options.pop(opt_name, None)
        
        def remove_options(self, opt_names: list) -> None:
            for opt_name in opt_names:
                self.remove_option(opt_name)
    
    
    class OptionControlButtons(Horizontal):
        def compose(self) -> ComposeResult:
            yield Button('Add Option ➕', id='add-option')
            yield Button("Refresh ↻", id='refresh')
            yield Button("Select All ✅ ", id='all')
            yield Button("Delete Selected 🗑️ ", id='delete')
            yield Button("↵ Back to log in", id='back-to-login')
            yield Button("Exit ❌", id='exit')
    
    
    class WarningScreen(Screen):
        def __init__(self, *instructions: Static | Pretty, **option_callbacks) -> None:
            super().__init__()
            self.instructions = instructions
            self.option_callbacks = option_callbacks
    
        def compose(self) -> ComposeResult:
            with VerticalScroll():
                for instruction in self.instructions:
                    yield instruction
            with Horizontal():
                for option_callback in self.option_callbacks:
                    yield Button(option_callback, id=option_callback)
    
        def on_button_pressed(self, event: Button.Pressed):
            event.stop()
    
            if (b_id := event.button.id) is not None and (
                callback := self.option_callbacks.get(b_id)
            ) is not None:
                callback()
    
            self.app.pop_screen()
    
    
    class SampleScreen(Screen):
        DEFAULT_CSS = """
        OptionControlButtons {
            height: auto;
            padding: 1;
        }
    
        VerticalScroll {
            padding: 1;
            height: auto;
        }
    
        SelectionList {
            padding: 1;
            border: solid $accent;
        }
        """
    
        def __init__(self) -> None:
            super().__init__(name="Sample Screen", id=None)
            self.server = FakeServer()
            self.options = SelectionList()
            self.options.border_title = f"Options"
    
        def on_mount(self) -> None:
            self.refresh_options()
    
        def refresh_options(self) -> None:
            all_options = self.server.options
            existing_options = [option for option in self.options._option_ids]
            
            removed = [option for option in existing_options if option not in all_options]
            for removed_option in removed:
                self.options.remove_option(removed_option)
            
            new_options = [option for option in all_options if option not in existing_options]
            for new_option in new_options:
                self.options.add_option(Selection(new_option, new_option, False, id=new_option))
    
        def delete_selected(self):
            if selected := self.options.selected:
                def yes_callback():
                    self.server.remove_options(selected)
                    self.refresh_options()
                self.app.push_screen(
                    WarningScreen(
                        Static("Following options will be deleted."),
                        Pretty(selected),
                        Static("Continue?"),
                        yes=yes_callback,
                        no=lambda: None,
                    )
                )
    
        def compose(self) -> ComposeResult:
            yield Header(show_clock=True)
            yield OptionControlButtons()
            with VerticalScroll():
                yield self.options
    
        def on_button_pressed(self, event: Button.Pressed):
            if event.button.id == "add-option":
                self.server.populate_option()
                event.stop()
                self.refresh_options()
            elif event.button.id == "delete":
                self.delete_selected()
                event.stop()
            elif event.button.id == 'refresh':
                self.refresh_options()
                event.stop()
            elif event.button.id == 'all':
                self.options.select_all()
                event.stop()
    
    class SampleLoginScreen(Screen):
        def compose(self) -> ComposeResult:
            yield Container(Button("LOG IN"))
    
        def on_button_pressed(self, _: Button.Pressed):
            self.app.push_screen(SampleScreen())
    
    
    class Sample(App):
        def on_mount(self) -> None:
            self.title = "Sample App"
            self.push_screen(SampleLoginScreen())
    
        def on_button_pressed(self, event: Button.Pressed) -> None:
            event.stop()
            if event.button.id == 'exit':
                self.exit()
            elif event.button.id == 'back-to-login':
                self.pop_screen()
    
    if __name__ == "__main__":
        app = Sample()
        app.run()

</details>

Here is the demo of the error with the code above:
![error_demo](https://github.com/Textualize/textual/assets/17974113/b9a46485-4188-4fa5-97a7-021479b009bc)

**Comment by davep** [comment]
This isn't an answer to the question, and that's quite a bit of code to wade through to be able to help (although it seems clear whatever is going on is happening around `SampleScreen.refresh_options`) so you might want to think about [working on a MRE to help narrow this down](https://github.com/Textualize/textual/discussions/3361), but an observation I wanted to make was about this line of code:

```python
existing_options = [option for option in self.options._option_ids]
```

`_option_ids` is a Textual internal (as is signified by the leading underscore). If you write code that relies on internals there's a very good chance it could stop working at some point in the future; either because the form of the data in that internal changes, or because it simply goes away.

**Reply by TomJGooding** [reply]
Unless I'm missing something, I can't see a public method to get the options to be fair! I suppose you could do something like:

```python
        existing_options: list[Selection] = [
            self.options.get_option_at_index(i)
            for i in range(self.options.option_count)
        ]
```

**Reply by YooSunYoung** [reply]
@davep Thanks for having a look! Sorry I tried to remove/simplify our tool but obviously it was not short enough...
I am aware of the hidden property you mentioned and it will not be part of our production code but thank you for reminding me!

I removed unnecessary part as much as possible again.

``` python
from textual.app import App, ComposeResult
from textual.containers import Horizontal, VerticalScroll
from textual.widgets import Button, SelectionList
from textual.widgets.selection_list import Selection


class OptionControlButtons(Horizontal):
    def compose(self) -> ComposeResult:
        yield Button('Add Option ➕', id='add-option')
        yield Button("Delete All 🗑️ ", id='delete-all')
        yield Button("Exit ❌", id='exit')


class Sample(App):
    def compose(self) -> ComposeResult:
        self.options = SelectionList()
        yield OptionControlButtons()
        with VerticalScroll():
            yield self.options

    def on_button_pressed(self, event: Button.Pressed) -> None:
        if event.button.id == 'exit':
            self.exit()
        elif event.button.id == "add-option":
            import uuid
            new_id = uuid.uuid4().hex
            self.options.add_option(
                Selection(new_id, new_id, False, id=new_id)
            )
            event.stop()
        elif event.button.id == "delete-all":
            self.options.select_all()
            if (selected := self.options.selected):
                for selected_opt in selected:
                    self.options.remove_option(selected_opt)
            event.stop()


if __name__ == "__main__":
    app = Sample()
    app.run()

```

And I think I found how to reproduce those errors.
It only happened if the `SelectionList` becomes empty, e.g. if you add options and then delete them all in this example.
Maybe I'm wrong... I still don't know the underlying cause...

**Reply by TomJGooding** [reply]
Thanks for reducing your example, that's really helpful! This looks like a possible bug with how `Option.highlighted` is calculated, I'm struggling to understand how/why it changes to a zero.

**Reply by TomJGooding** [reply]
It looks like `OptionList` was designed to always have a highlighted option based on the validation?

https://github.com/Textualize/textual/blob/e68e6e37b546c8c7d760bc2ffd7fa8cfa3064907/src/textual/widgets/_option_list.py#L987-L992

But if you start with an empty `OptionList` and then add options, the validation is never called as I assume `highlighted` isn't assigned. But it definitely is assigned when an option is removed.

I'm not sure yet exactly how it causes the error in this example, but just sharing what I've found so far.

**Reply by davep** [reply]
I *think* we can narrow this down to this:

```python
from textual.app import App, ComposeResult
from textual.widgets import SelectionList
from textual.widgets.selection_list import Selection

class SelectionListError(App[None]):

    def compose(self) -> ComposeResult:
        yield SelectionList[str](Selection("Test", "test", id="test"))

    def on_mount(self) -> None:
        self.query_one(SelectionList).remove_option("test")

if __name__ == "__main__":
    SelectionListError().run()
```

The issue *seems* to be something going wonky with IDs (but this is a I'm-on-holiday-so-not-really-here driveby check so take it with a pinch of salt for now).

**Reply by davep** [reply]
Actually, scratch that, I suspect that the ID list won't have had a chance to be built by that point so I'm throwing myself off there.

**Reply by TomJGooding** [reply]
Stop checking notifications and enjoy your holiday!

Was I doing _such_ a bad job at trying to debug this you had to swoop in? ;-)

**Reply by davep** [reply]
You were doing a far better job than I was, 'cos I totally confused myself.

PS: I think the bug might be in `SelectionList._on_option_list_option_highlighted`; at least it might point to the cause. If I add a check that `self.highlighted is not None` it removes the issue:

```python
        event.stop()
        if self.highlighted is not None:
            self.post_message(self.SelectionHighlighted(self, event.option_index))
```

If that isn't the fix, I suspect that points to the underlying cause (but I suspect this is the thing to be doing).

**Reply by TomJGooding** [reply]
Right, that's what was confusing me since this shouldn't be None based on when `OptionHighlighted` is posted. Especially when nothing is actually highlighted in the example above!

**Reply by davep** [reply]
Yeah, that bit is very curious.

**Reply by TomJGooding** [reply, marked as answer]
While checking back on my outstanding issues, I found #3560 which I raised off the back off this discussion. I _think_ this problem is resolved in the latest version of Textual, though I confess my memory is a bit hazy on the specifics!

**Reply by YooSunYoung** [reply]
Thank you!
I also tried again and it worked as you said so I could update my tool in our package : D

**Comment by TomJGooding** [comment]
I suspect the problem is that your `FakeServer` updates and Textual interface are getting out of sync - perhaps the [Worker API ](https://textual.textualize.io/guide/workers/) would help here.

**Reply by YooSunYoung** [reply]
Thank you for the advice, it is indeed what we needed, but the error still occurs even though I don't use the `FakeServer`... 
I posted a shorter example in another thread: https://github.com/Textualize/textual/discussions/3559#discussioncomment-7324184

**Comment by YooSunYoung** [comment]
I am currently using a workaround to replace ``SelectionList`` with a new ``SelectionList`` instance when it's refreshed.

This seems like to solve the problem but I am not sure if it is a good practice to destroy and create a widget instance so frequently....
